### PR TITLE
MS-690 Remote Headless & Non-Headless Acceptance Testing within Docker

### PR DESCRIPTION
**Pull request by jjrobothamho**
This Pull request will allow you to run remote acceptance tests using Google Chrome within docker containers.

Tests can only be run in headless mode (the browser runs in the background)

### CODE REVIEW

**Comment by sulthan-ahmed** [comment]
Rebase?

File: acceptance-test/bootstrap/browser.js [hunk 1]
```diff
@@ -1,11 +1,61 @@
...
+ *
+ * This is needed so we can extract the browser websocket endpoint from the body
+ * of the request
+ *
+ * @see getBrowserWebSocketEndpoint
+ *
+ * @returns {Object} - the response from the browser container
+ */
+const getBrowserContainerResponse = async() => {
+    // Workaround for https://github.com/GoogleChrome/puppeteer/issues/2242
```
Comment (sulthan-ahmed) [inline acceptance-test/bootstrap/browser.js]: Can you explain this workaround? Saves a dev from clicking on the link

File: browsers/chrome/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+# Install Google Chrome
+RUN curl https://dl-ssl.google.com/linux/linux_signing_key.pub | apt-key add - && \
+  echo 'deb [arch=amd64] http://dl.google.com/linux/chrome/deb stable main' >> /etc/apt/sources.list.d/google-chrome.list && \
+  apt-get update && \
+  apt-get install -y google-chrome-unstable --no-install-recommends && \
+  rm -fr /var/lib/apt/lists/* && \
+  apt-get purge --auto-remove -y curl && \
+  rm -fr /src/*.deb
+
```
Comment (sulthan-ahmed) [inline browsers/chrome/Dockerfile]: Can you explain the below in comments?

File: browsers/chrome/entrypoint.sh [hunk 3]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
```
Comment (sulthan-ahmed) [inline browsers/chrome/entrypoint.sh]: Can you explain this file is?

File: acceptance-test/user-pathways/upload-file/upload-file.test.js [hunk 4]
```diff
@@ -19,7 +19,7 @@ const {
 describe('Upload File(s)', () => {
 
     it('upload 1 document', async() => {
-        const { browser, page, hostIP: APP_CONTAINER_HOST } = await bootstrap.buildBrowser();
+        const { page, hostIP: APP_CONTAINER_HOST } = await bootstrap.buildBrowser();
 
         const { env: { PORT } } = process;
```
Comment (sulthan-ahmed) [inline acceptance-test/user-pathways/upload-file/upload-file.test.js]: O I missed this from before, what's going on here? I know it's related to your other PR

**Comment by sulthan-ahmed** [comment]
Tested that it runs and it does, good job, just some slight suggestions for commenting

File: acceptance-test/user-pathways/upload-file/upload-file.test.js (hunk 4 above)
Comment (jjrobothamho) [reply acceptance-test/user-pathways/upload-file/upload-file.test.js]: Do you mean the destructuring of the returned object?
This is another way of writing:
const myBrowser = await bootstrap.buildBrowser();
const browser = myBrowser.browser;
const page = myBrowser. page;
const APP_CONTAINER_HOST = myBrowser. hostIP;

or

What the line achieves?
Uses puppeteer to create a connection with the container in which the browser lies and returns the resulting browser object (The browser which puppeteer controls), the page object (the page which is now open as a result of the browser opening) and the hostIP for the browser being run in Chrome Dev tools for use within the URL constructed later on in the script.

Let me know if that doesn't answer the question you had.

Comment (sulthan-ahmed) [reply acceptance-test/user-pathways/upload-file/upload-file.test.js]: it was more this one
`const { env: { PORT } } = process;` and what it achieved.  Is it just adding `Process.env.PORT` to `process`?

**Comment by sulthan-ahmed** [comment]
LGTM

**Pull request merged**